akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/SliceRangeShardAllocationStrategySpec.scala

```diff
@@ -0,0 +1,519 @@
+      val member1 = newUpMember("127.0.0.1", upNbr = 1, version = Version("1.0.0"))
+      val member2 = newUpMember("127.0.0.2", upNbr = 2, version = Version("1.0.1"))
+      val member3 = newUpMember("127.0.0.3", upNbr = 3, version = Version("1.0.0"))
+
+      val allocationStrategy = createAllocationStrategy(Vector(member1, member2, member3))
+      val allocations = Map(
+        regionA -> (0 until 300).map(_.toString).toVector,
+        regionB -> (300 until 600).map(_.toString).toVector,
+        regionC -> Vector.empty)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set.empty[String])
+    }
+
+    "not rebalance when regions are unreachable" in {
+      val allocationStrategy =
+        new SliceRangeShardAllocationStrategy(10, 0.1) {
+          override protected def clusterState: CurrentClusterState =
+            CurrentClusterState(SortedSet(memberA, memberB, memberC), unreachable = Set(memberB))
+
+          override protected def selfMember: Member = memberB
+        }
+      val allocations = Map(
+        regionA -> (0 until 300).map(_.toString).toVector,
+        regionB -> (300 until 600).map(_.toString).toVector,
+        regionC -> Vector.empty)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set.empty[String])
+    }
+
+    "not rebalance when members are joining dc" in {
+      val allocationStrategy =
+        new SliceRangeShardAllocationStrategy(10, 0.1) {
+
+          val member1 = newUpMember("127.0.0.1", upNbr = 1)
+          val member2 =
+            Member(
+              UniqueAddress(Address("akka", "myapp", "127.0.0.2", 252525), 1L),
+              Set(ClusterSettings.DcRolePrefix + ClusterSettings.DefaultDataCenter),
+              member1.appVersion).copyUp(2)
+          val member3 = newUpMember("127.0.0.3", upNbr = 3)
+
+          override protected def clusterState: CurrentClusterState =
+            CurrentClusterState(SortedSet(member1, member2, member3), unreachable = Set.empty)
+          override protected def selfMember: Member = member2
+        }
+      val allocations = Map(
+        regionA -> (0 until 300).map(_.toString).toVector,
+        regionB -> (300 until 600).map(_.toString).toVector,
+        regionC -> Vector.empty)
+      allocationStrategy.rebalance(allocations, Set.empty).futureValue should ===(Set.empty[String])
+
+    }
+
+  }
+
+  "ShardBySliceMessageExtractor" must {
+    "extract slice as shardId" in {
+      val persistence = Persistence(system)
+      val extractor = new ShardBySliceMessageExtractor("TestEntity", persistence)
+      val shardId = extractor.shardId("abc")
+      val slice = shardId.toInt
+      slice should ===(persistence.sliceForPersistenceId("TestEntity|abc"))
+    }
+  }
+
+  // These are not real tests, but can be useful for exploring the algorithm and tuning
+  "SliceRangeShardAllocationStrategy simulations" must {
+
+    "try distributions" ignore {
```

Comment:
Ok, so for the numbers we've been testing with (around 25 nodes and 8 slice ranges) it's a decent improvement, but still over 3x the optimal number of connections. And for higher number of nodes it's easier to get affinity.

Reply:
Let's say a connection pool of max 20, then 73*20/8=274 connections per db should still be fine.
I'd say that it's more important for the larger clusters, so good that it looks better for that.

Reply:
One idea that I have been pondering is if we should have a pre-allocation phase where it would allocate all shards in order 0-1023. Then it would be near perfect affinity. Would have to be triggered from the outside of the allocation strategy itself, in the coordinator, or by something asking the coordinator for shard homes. Could be triggered when reaching `min-nr-of-members`.

However, it would only help for the initial allocation and not for rebalance.

Reply:
I made a small change that had a significant improvement for the smaller clusters. I adjust how far to look for neighbors depending on optimal range size, previously it was hardcoded to 10. bc9f86d90a

New results:
```
total of 46 connections from 20 nodes to 8 backend ranges, reduction by 114
total of 67 connections from 20 nodes to 16 backend ranges, reduction by 253

total of 84 connections from 50 nodes to 8 backend ranges, reduction by 316
total of 98 connections from 50 nodes to 16 backend ranges, reduction by 702

total of 124 connections from 100 nodes to 8 backend ranges, reduction by 676
total of 138 connections from 100 nodes to 16 backend ranges, reduction by 1462
```

Reply:
Cool. Makes sense that it has more opportunity to coalesce with looking further.